com.unity.ml-agents/Runtime/DemonstrationRecorder.cs

```diff
@@ -65,31 +93,63 @@ public static string SanitizeName(string demoName, int maxNameLength)
         }
 
         /// <summary>
-        /// Forwards AgentInfo to Demonstration Store.
+        /// Gets a unique path for the demonstrationName in the demonstrationDirectory.
         /// </summary>
-        public void WriteExperience(AgentInfo info, List<ISensor> sensors)
+        /// <param name="fileSystem"></param>
+        /// <param name="demonstrationDirectory"></param>
+        /// <param name="demonstrationName"></param>
+        /// <returns></returns>
+        internal static string MakeDemonstrationFilePath(
+            IFileSystem fileSystem, string demonstrationDirectory, string demonstrationName
+        )
         {
-            m_DemoStore.Record(info, sensors);
+            // Create the directory if it doesn't already exist
+            if (!fileSystem.Directory.Exists(demonstrationDirectory))
+            {
+                fileSystem.Directory.CreateDirectory(demonstrationDirectory);
+            }
+
+            var literalName = demonstrationName;
+            var filePath = Path.Combine(demonstrationDirectory, literalName + k_ExtensionType);
+            var uniqueNameCounter = 0;
+            while (fileSystem.File.Exists(filePath))
+            {
+                // TODO should we use a timestamp instead of a counter here? This loops an increasing number of times
+                // as the number of demos increases.
+                literalName = demonstrationName + "_" + uniqueNameCounter;
+                filePath = Path.Combine(demonstrationDirectory, literalName + k_ExtensionType);
+                uniqueNameCounter++;
+            }
+
+            return filePath;
         }
 
+        /// <summary>
+        /// Close the DemonstrationStore and remove it from the Agent.
+        /// Has no effect if the DemonstrationStore is already closed (or wasn't opened)
+        /// </summary>
         public void Close()
         {
             if (m_DemoStore != null)
             {
+                var agent = GetComponent<Agent>();
+                if (agent != null)
+                {
+                    agent.DemonstrationStores.Remove(m_DemoStore);
+                }
+
                 m_DemoStore.Close();
                 m_DemoStore = null;
             }
         }
 
         /// <summary>
-        /// Closes Demonstration store.
+        /// Clean up the DemonstrationStore when shutting down.
         /// </summary>
         void OnApplicationQuit()
```

Comment:
Why not onDestroy ?

Reply:
That's how it was before :) But yeah, sounds like a better place for it.